### ISSUE TITLE
Does Origen Testers support V93K test bypass?

### ISSUE THREAD

**Issue by info-rchitect** [Origen-SDK/origen_testers#79]
The test is in the flow but turns yellow.   thx

**Comment by ginty** [comment]
Not currently, but have thought before that we should have one.

I think the API should be:

~~~ruby
test :my_test, disabled: true
~~~

And that would resolve to setting the bypass flag for the given test on 93K, and setting the opcode to nop on IG-XL.

**Comment by info-rchitect** [comment]
OK thanks @ginty.  I do think we might be able to handle this one as well but until then, do you think that just wrapping a test with if_enable would work?

~~~ruby
if_enable :disabled_tests do
  func :test1
end
~~~

What is the actual difference here?

**Comment by ginty** [comment]
Yeah that would work fine, and now you mention it, that is probably why we haven't actually added the ability to generate a bypass before now.

If at generation time you know that there is value in adding a test but you don't want it enabled by default, then when wouldn't you use the enable API like you have shown since it gives you a top-level control to be able to turn it on when you want to.

For a bypass, it's implemented right down at the test level, so to turn it on you need to drill down to that test in the flow and then un-check the bypass switch.

I think bypass/nop is really more intended to be used when debugging within the ATE toolset - e.g. if a test doesn't work then just disable for now and come back it later.
Generating bypassed tests in the first place doesn't make so much sense though.

Perhaps there is an argument that if you want to make what comes out of your program generator always be able to pass, then it may make sense to be able to quickly implement the same switch at the source level. Though as you note adding `if_enable: :blah` is just as easy as typing `disabled: true`, and ultimately has the same outcome.

**Comment by info-rchitect** [comment]
@ginty thx for the explanation.  Our flows are heavily looped and product agnostic such that 14 lines of code in the flow file generates > 500 tests.  If silicon debug identifies 38 tests, that exists in different locations in the loops, that are not stable and need to be disabled or removed, we didn't want to have to make edits to our flow file.  So we slurp in a text file:

~~~shell
test_expression, action
pcie_register_read_vmin, disable
pcie_tx_rx_detect_vmin, remove
/5000/, disable
/phy/, remove
~~~

Removing a test is easy, don't create it in the first place but disabling a test is why this issue arose.  In the end, we want to be able to do unit level testing of each test modules tests as so:

~~~ruby
[3] pry(#<Test::Interface>)> Origen.top_level.test_modules(:pcie).tests.map { |category, tests| "Tests #{category}: #{tests.size}" }
=> ["Tests defined: 108", "Tests enabled: 94", "Tests disabled: 1", "Tests removed: 13"]
~~~

**Comment by arkohler** [comment]
@ginty, as a use case for disabled tests consider test nodes used for debug.
I have a set of tests where my patterns read pass/fail information out of the part which I compare and use to bin.  The part has some detailed diagnostic info inside registers that I can go read to help debug particular fails.  I don't want to collect the diagnostic info on all fails in production, but I also don't want to have to set up the test node from scratch every time I sit down for debug.  So I set up the test node in the flow, but configured to skip unless I explicitly enable it.
I agree that the if_enable keywords can produce much the same effect.

**Comment by info-rchitect** [comment]
@ginty We are using the if_enable workaround but the general feedback is that a native V93K bypass is best.  We can tackle the PR if you have a little background on where  the code resides.

**Comment by ginty** [comment]
Probably you can add :bypass to the existing conditions in ATP:
https://github.com/Origen-SDK/atp/blob/master/lib/atp/flow.rb#L9

That should then allow you pass in `bypass: true` to a given test and also give you the block form:

~~~ruby
bypass true do
  # All these guys should be bypassed by default
end
~~~

Then you need to make a handler for it in the flow generator, something like this:

Initialize a bypass stack here: https://github.com/Origen-SDK/origen_testers/blob/master/lib/origen_testers/smartest_based_tester/base/flow.rb#L92

~~~ruby
@bypass = []
end

Then create the handler here: https://github.com/Origen-SDK/origen_testers/blob/master/lib/origen_testers/smartest_based_tester/base/flow.rb#L206

~~~ruby
def on_bypass(node)
  value, *children = *node
  # Store the value of bypass in a stack, a stack is used so that nested bypass is supported
  @bypass << value
  process_all(children)   # Processes all the test and other nodes wrapped by this bypass condition
  @bypass.pop
end
~~~

Then modify the test node handler to apply the bypass if required, https://github.com/Origen-SDK/origen_testers/blob/master/lib/origen_testers/smartest_based_tester/base/flow.rb#L111

~~~ruby
if @bypass.last
  # Add the bypass flag
end
~~~

I would be surprised if that just works as written, but its not far off what's required.

If you want to understand wtf is going on with the AST nodes better, spend some time reading these pages:

http://www.rubydoc.info/gems/ast/AST/Node

http://www.rubydoc.info/gems/ast/AST/Processor/Mixin

**Comment by arkohler** [comment]
Circling back on an old thread.
We discovered that origen does indeed support the bypass flag natively for tests, as long as it is passed as an option to test_suites.add().

**Comment by ginty** [comment]
Thanks for bumping this thread @arkohler as I was already thinking that I would need to go and look for it since it relates to #87. 

Thanks also for noting that about the test suites, I didn't realize that.
We won't take that feature away, but I would consider `test_suites.add(:blah, bypass: true)` to be a V93K specific API (and actually probably SMT7 specific...)

It is interesting that SMT7 treats that as a test suite/instance parameter whereas IG-XL treats it as a flow-level attribute.
I think for supporting this officially in Origen we will also treat it as a flow level attribute.
I would also note that I think SMT8 must treat this as a flow-level attribute too since it has similar architecture to IG-XL in the sense that you no longer have one test suite definition per flow line. Instead, you can have a single test suite definition and call it multiple times in the flow and you would not necessarily want to always bypass them all.

Anyway, I propose to add the following which will be supported across the board (IG-XL, SMT7, SMT8):

~~~ruby
# A single test, bypassed by default
test :blah, bypass: true

# All tests wrapped in this block will be bypassed by default
bypass do
  test :blah
  test :another_blah
end

# A group that is bypassed by default, on supporting platforms the flag will be applied to the group,
# otherwise it is equivalent to the above and all contained tests will have their bypass flag set
group :blah, bypass: true do
   test :blah
   test :another_blah
end
~~~

I've assigned this to myself, athough I'm not placing high priority on it since it seems that this and #87 have available workarounds for the time being.
